numpy/_core/src/multiarray/multiarraymodule.h

```diff
@@ -1,26 +1,114 @@
 #ifndef NUMPY_CORE_SRC_MULTIARRAY_MULTIARRAYMODULE_H_
 #define NUMPY_CORE_SRC_MULTIARRAY_MULTIARRAYMODULE_H_
 
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_current_allocator;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_function;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_struct;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_priority;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_interface;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_wrap;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_finalize;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_implementation;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_axis1;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_axis2;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_like;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_numpy;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_where;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_convert;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_preserve;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_convert_if_no_array;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_cpu;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_dtype;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str_array_err_msg_substr;
-NPY_VISIBILITY_HIDDEN extern PyObject * npy_ma_str___dlpack__;
+/*
+ * A struct storing thread-unsafe global state for the _multiarray_umath
+ * module. We should refactor so the global state is thread-safe,
+ * e.g. by adding locking.
+ */
+typedef struct npy_thread_unsafe_state_struct {
+    /*
+     * Cached references to objects obtained via an import. All of these are
+     * can be initialized at any time by npy_cache_import.
+     *
+     * Currently these are not initialized in a thread-safe manner but the
+     * failure mode is a reference leak for references to imported immortal
+     * modules so it will never lead to a crash unless users are doing something
+     * janky that we don't support like reloading.
+     *
+     * TODO: maybe make each entry a struct that looks like:
+     *
+     *      struct {
+     *          atomic_int initialized;
+     *          PyObject *value;
+     *      }
```

Comment:
You need to use at least one atomic load otherwise it's possible that the compiler might reorder things like:

```
// original code
if (flag) {
    do_some_work();
    do_more_work();
}

// after optimization
if (flag) {
    do_some_work();
}
if (flag) {
    do_more_work()
}
```

because data races are UB. In this case it would just lead to a memory leak. For the argparse cache I think it could cause two threads to simultaneously run `initialize_keywords`, corrupting the cache.

Reply:
Yeah, I suspected that much, was just wondering if the `atomic_int` is strictly necessary (or there is a way to do an `atomic_load(value)`, I guess).

Reply:
Ah no it's not necessary, I can do it with an atomic load instead.